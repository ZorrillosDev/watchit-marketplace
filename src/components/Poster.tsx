// mui imports
import {
  Avatar,
  AvatarGroup,
  Box, CardHeader,
  Grid,
  styled,
  Typography,
  IconButton, CardMedia,
  CardContent, Card, CardProps, CardHeaderProps, CardContentProps, Menu, MenuItem, Tooltip
} from '@mui/material'
import React, { FC } from 'react'
import { MoreHoriz } from '@mui/icons-material'
import HeartCounter from '@components/HeartCounter'
import TruncatedTypography from '@components/TruncatedTypography'

// ===========================|| POSTER ||=========================== //

export interface PosterProps {
  creator: {
    username: string
    profileUrl: string
  }
  owner: {
    username: string
    profileUrl: string
  }
  posterUrl: string
  price: number
  title: string
  description: string
  rate: number
  isFavorite: boolean
}

const Poster: FC<PosterProps & {showDetails: boolean}> = (props): JSX.Element => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)
  const open = Boolean(anchorEl)

  const handleClick = (event: React.MouseEvent<HTMLElement>): void => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = (): void => {
    setAnchorEl(null)
  }

  return (
    <PosterWrapper showDetails={props.showDetails}>
      {props.showDetails && (
        <>
          <PosterHeader
            avatar={
              <AvatarGroup max={2}>
                <Tooltip title={`Owner: ${props.owner.username}`}>
                  <Avatar alt={`${props.owner.username} profile image`} src={`${props.owner.profileUrl}`} sx={{ width: 24, height: 24 }} />
                </Tooltip>
                <Tooltip title={`Creator: ${props.creator.username}`}>
                  <Avatar alt={`${props.creator.username} profile image`} src={`${props.creator.profileUrl}`} sx={{ width: 24, height: 24 }} />
                </Tooltip>
              </AvatarGroup>
            }
            action={
              <IconButton aria-label='settings' onClick={handleClick}>
                <MoreHoriz />
              </IconButton>
            }
          />
          <Menu
            anchorEl={anchorEl}
            open={open}
            onClose={handleClose}
            transformOrigin={{
              vertical: 'top',
              horizontal: 'right'
            }}
            anchorOrigin={{
              vertical: 'bottom',
              horizontal: 'right'
            }}
          >
            <MenuItem onClick={handleClose}>
              Purchase
            </MenuItem>
            <MenuItem onClick={handleClose}>
              Share
            </MenuItem>
            <MenuItem onClick={handleClose}>
              Report
            </MenuItem>
          </Menu>
        </>
      )}
      <PosterMediaContent showDetails={props.showDetails}>
        <CardMedia
          component='img' image={`${props.posterUrl}`} alt={`${props.title}`}
          sx={{ pointerEvents: 'none', userSelect: 'none' }}
        />
      </PosterMediaContent>
      {props.showDetails && (
        <PosterContent>
          <Grid container spacing={1}>
            <Grid item xs={12}>
              <TruncatedTypography variant='body2' color='primary.dark' fontWeight={600} lines={1} sx={{ mt: 0 }}>
                {`${props.title}`}
              </TruncatedTypography>
            </Grid>
            <Grid item xs={6}>
              <Box display='flex' alignItems='center'>
                <Typography variant='body1' display='inline' color='text.primary' fontWeight='bold'>{props.price} ETH</Typography>
              </Box>
            </Grid>
            <Grid item display='flex' alignItems='center' justifyContent='flex-end' xs={6}>
              <HeartCounter count={props.rate} favorite={props.isFavorite} />
            </Grid>
          </Grid>
        </PosterContent>
      )}
    </PosterWrapper>
  )
}

export default Poster

export const PosterWrapper = styled(Card, {
  shouldForwardProp: (prop) => prop !== 'showDetails'
})<CardProps & {showDetails: boolean}>(({ theme, showDetails }) => ({
  position: 'relative',
  backgroundColor: theme.palette.background.paper,
  boxShadow: showDetails ? '0 3px 6px rgba(0,0,0,0.16)' : '',
  cursor: 'pointer',
  width: '100%',
  height: '100%',
  transition: 'all 0.3s ease-in-out',
  '&:hover': {
    transform: 'translateY(-1rem)'
  }
}))

export const PosterHeader = styled(CardHeader)<CardHeaderProps>(() => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  width: '100%',
  padding: '0.25rem 0.5rem'
}))

export const PosterContent = styled(CardContent)<CardContentProps>(() => ({
  padding: '0.5rem !important',
  svg: {
    width: '0.9rem',
    height: '0.9rem'
  }
}))

export const PosterMediaContent = styled(CardContent, {
  shouldForwardProp: (prop) => prop !== 'showDetails'
})<CardContentProps & {showDetails: boolean}>(({ showDetails }) => ({
  padding: '0.5rem !important',
  width: '100%',
  height: '100%',
  maxHeight: showDetails ? 'calc(100% - 7rem)' : 'auto',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  img: {
    height: 'auto',
    maxHeight: '100%',
    maxWidth: '100%',
    borderRadius: '6px',
    width: showDetails ? 'auto !important' : '100%'
  }
}))
