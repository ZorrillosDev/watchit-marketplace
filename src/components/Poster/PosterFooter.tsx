// REACT IMPORTS
import React, { FC } from 'react';

// MUI IMPORTS
import { Box, Grid, styled, Typography, CardContent, CardContentProps } from '@mui/material';

// PROJECT IMPORTS
import HeartCounter from '@components/HeartCounter';
import TruncatedTypography from '@components/TruncatedTypography';

/* eslint-disable  @typescript-eslint/strict-boolean-expressions */

// ===========================|| POSTER - FOOTER ||=========================== //

export interface PosterFooterProps {
  name: string
  price: number
  isEthPrice?: boolean
  rating?: number
}

export const PosterFooter: FC<PosterFooterProps> = (props): JSX.Element => {
  return (
    <PosterFooterWrapper>
      <Grid container spacing={1}>
        <Grid item xs={12}>
          <TruncatedTypography
            variant='body2'
            color='primary.dark'
            fontWeight={600} lines={1}
            sx={{ mt: 0 }}
          >
            {`${props.name}`}
          </TruncatedTypography>
        </Grid>
        <Grid item xs={6}>
          <Box display='flex' alignItems='center'>
            <Typography
              variant='body1'
              display='inline'
              color='text.primary'
              fontWeight='bold'
            >
              {props.price} ETH
            </Typography>
          </Box>
        </Grid>
        <Grid item display='flex' alignItems='center' justifyContent='flex-end' xs={6}>
          <HeartCounter count={props.rating ?? 0} />
        </Grid>
      </Grid>
    </PosterFooterWrapper>
  );
};

export const PosterFooterWrapper = styled(CardContent)<CardContentProps>(() => ({
  padding: '0.1rem 0.9rem !important',
  svg: {
    width: '0.9rem',
    height: '0.9rem',
  },
}));
