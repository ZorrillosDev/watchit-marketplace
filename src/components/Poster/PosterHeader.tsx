// REACT IMPORTS
import React, { FC } from 'react'

// MUI IMPORTS
import {
  Avatar, AvatarGroup, styled,
  IconButton, Menu, MenuItem, Tooltip, Box, BoxProps
} from '@mui/material'

// PROJECT IMPORTS
import { MoreHoriz } from '@components/Icons'

// ===========================|| POSTER HEADER ||=========================== //

export interface PosterHeaderProps {
  creator: {
    username: string
    profileUrl: string
  }
  owner: {
    username: string
    profileUrl: string
  }
}

export const PosterHeader: FC<PosterHeaderProps> = (props): JSX.Element => {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null)
  const open = Boolean(anchorEl)

  const handleClick = (event: React.MouseEvent<HTMLElement>): void => {
    setAnchorEl(event.currentTarget)
  }

  const handleClose = (): void => {
    setAnchorEl(null)
  }

  return (
    <>
      <PosterHeaderWrapper>
        <AvatarGroup max={2}>
          <Tooltip title={`Owner: ${props.owner.username}`}>
            <Avatar
              alt={`${props.owner.username} profile image`}
              src={`${props.owner.profileUrl}`} sx={{ width: 24, height: 24 }}
            />
          </Tooltip>
          <Tooltip title={`Creator: ${props.creator.username}`}>
            <Avatar
              alt={`${props.creator.username} profile image`}
              src={`${props.creator.profileUrl}`} sx={{ width: 24, height: 24 }}
            />
          </Tooltip>
        </AvatarGroup>
        <IconButton aria-label='settings' onClick={handleClick} size='small'>
          <MoreHoriz />
        </IconButton>
        <Menu
          anchorEl={anchorEl}
          open={open}
          onClose={handleClose}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'right'
          }}
          anchorOrigin={{
            vertical: 'bottom',
            horizontal: 'right'
          }}
        >
          <MenuItem onClick={handleClose}>
            Purchase
          </MenuItem>
          <MenuItem onClick={handleClose}>
            Share
          </MenuItem>
          <MenuItem onClick={handleClose}>
            Report
          </MenuItem>
        </Menu>
      </PosterHeaderWrapper>
    </>
  )
}

export const PosterHeaderWrapper = styled(Box)<BoxProps>(() => ({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  width: '100%',
  padding: '0.4rem 0.9rem 0.2rem'
}))
