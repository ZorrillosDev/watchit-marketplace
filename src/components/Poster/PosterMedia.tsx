// REACT IMPORTS
import React, { FC } from 'react'

// MUI IMPORTS
import { styled, CardMedia, CardContent, CardContentProps, Theme } from '@mui/material'
import { SxProps } from '@mui/system'

// ===========================|| POSTER - MEDIA ||=========================== //

export interface PosterMediaProps {
  image: string
  name: string
  sx?: SxProps<Theme>
}

export const PosterMedia: FC<PosterMediaProps> = (props): JSX.Element => {
  return (
    <PosterMediaContent sx={props.sx}>
      <CardMedia
        component='img' image={`${props.image}`} alt={`${props.name}`}
        sx={{ pointerEvents: 'none', userSelect: 'none' }}
      />
    </PosterMediaContent>
  )
}

export const PosterMediaContent = styled(CardContent)<CardContentProps>(() => ({
  padding: '0.5rem !important',
  width: '100%',
  height: '100%',
  maxHeight: 'calc(100% - 6.65rem)',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  img: {
    height: 'auto',
    maxHeight: '100%',
    maxWidth: '100%',
    borderRadius: '6px',
    width: '65%'
  }
}))
