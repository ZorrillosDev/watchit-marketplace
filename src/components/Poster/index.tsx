// ===========================|| POSTER EXPORTS ||=========================== //

export { Poster as default } from '@components/Poster/Poster';
export { PosterWrapper } from '@components/Poster/Poster';
export * from '@components/Poster/PosterFooter';
export * from '@components/Poster/PosterHeader';
export * from '@components/Poster/PosterMedia';
