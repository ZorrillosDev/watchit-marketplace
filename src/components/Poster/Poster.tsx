// REACT IMPORTS
import React, { FC } from 'react'

// MUI IMPORTS
import { styled, Card, CardProps } from '@mui/material'

// PROJECT IMPORTS
import { PosterMediaProps, PosterMedia } from '@components/Poster/PosterMedia'

// ===========================|| POSTER ||=========================== //

export const Poster: FC<PosterMediaProps> = (props): JSX.Element => {
  return (
    <PosterWrapper sx={{ border: 'none' }}>
      <PosterMedia {...props} sx={{ maxHeight: 'none', img: { width: '100%' } }} />
    </PosterWrapper>
  )
}

export const PosterWrapper = styled(Card)<CardProps>(() => ({
  position: 'relative',
  backgroundColor: 'transparent',
  border: '1px solid #eee',
  boxShadow: 'none',
  cursor: 'pointer',
  width: '100%',
  height: '100%',
  transition: 'all 0.3s ease-in-out',
  '&:hover': {
    transform: 'translateY(-1rem)'
  }
}))
