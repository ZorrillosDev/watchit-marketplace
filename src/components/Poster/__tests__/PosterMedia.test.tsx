import React from 'react';
import { shallow } from 'enzyme';
import { PosterMedia } from '@components/Poster';

describe('<PosterMedia />', () => {
  it('should render', () => {
    const component = shallow(<PosterMedia name='test' image='test' />);

    expect(component).toMatchSnapshot();
  });
});
