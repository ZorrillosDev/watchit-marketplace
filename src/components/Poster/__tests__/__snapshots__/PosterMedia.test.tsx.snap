// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PosterMedia /> should render 1`] = `ShallowWrapper {}`;
